networkx/algorithms/dag.py

```diff
@@ -170,10 +180,13 @@ def topological_sort(G):
 
     indegree_map = {v: d for v, d in G.in_degree() if d > 0}
     # These nodes have zero indegree and ready to be returned.
-    zero_indegree = [v for v, d in G.in_degree() if d == 0]
+    zero_indegree = deque(v for v, d in G.in_degree() if d == 0)
+
+    generation = 0
+    last_node_in_generation = zero_indegree[-1] if zero_indegree else None
```

Comment:
setting last_node_in_generation to None could be an issue if None was a valid node. However, this led me to discover another existing bug, which makes this decision irrelevant.

None if a valid node allowed in a graph. However if you try to topo_sort a graph with None in it (on main branch) it will fail, because when it looks for edges of node=None, it will instead return all edges in the graph, instead of the None-node. 

I think 1 of the following actions should probably be taken:
1) None shouldn't be allowed as a node in a graph (seems like overkill potentially)
2) The edges method should do something like this instead of current. With this suggestion None can be used as a regular node.
```
__empty = object()
def edges(self, nbunch=__empty, data=False, default=None):
    ...
```
3) don't allow topo_sort when None exists as a node.

Reply:
@rossbar @dschult  ^

Reply:
Nodes can be "any hashable except the object `None`"

Reply:
Makes sense. Has it ever been discussed to explicitly forbid `None`, rather than by convention? Since it seems as of now nothing stops you from using it as a node, and can lead to ambiguous errors in algorithms. In any case this is a discussion for another thread. If you think it's worth pursuing I can open a separate issue for it.

Reply:
+1 for opening a new [discussion topic](https://github.com/networkx/networkx/discussions) so we can discuss this issue separately, otherwise it we'll be juggling too many separate convos in this PR :).

Given that `None` can't be a node in a graph, the way `last_node_in_generation` is handled here seems fine since it's really just a state variable and is never added to a Graph object or returned from the function AFAICT.